### PR TITLE
feat: When a PR is associated with an Issue and uses close semantics, the remaining score will be automatically rewarded to this PR #211

### LINKED ISSUE #211: When a PR is associated with an Issue and uses close semantics, the remaining score will be automatically rewarded to this PR
### Is your feature request related to a Issue or PR(Optional)
None 

### Do you have any ideas on how to implement it?(Optional)

When a PR is associated with an Issue and uses close semantics, the remaining score will be automatically rewarded to this PR.

### PR DESCRIPTION
**Pull request by whileTrue0x90**
Issue Number: close #211

### CODE REVIEW

File: src/services/challenge-pull/index.ts [hunk 1]
```diff
@@ -385,4 +385,63 @@ export default class ChallengePullService implements IChallengePullService {
       message: ChallengePullMessage.Rewarded,
     };
   }
+
+  private async getCurrentIssueLeftScore(issueNumber: number, number: number) {
```
Comment (hi-rustin) [inline src/services/challenge-pull/index.ts:389]: No need add this new function.

File: src/services/challenge-pull/index.ts [hunk 2]
```diff
@@ -385,4 +385,63 @@ export default class ChallengePullService implements IChallengePullService {
...
+    return await this.scoreRepository.getCurrentIssueLeftScore(
+      issueNumber,
+      number
+    );
+  }
+
+  private async getIssueIdByIssueNumber(
```
Comment (hi-rustin) [inline src/services/challenge-pull/index.ts:396]: ditto

**Review by hi-rustin — CHANGES_REQUESTED** [review]

**Comment by hi-rustin** [comment]
Please change your PR title.

File: src/services/challenge-pull/index.ts [hunk 3]
```diff
@@ -385,4 +385,39 @@ export default class ChallengePullService implements IChallengePullService {
...
+    const closeIndex = pullPayload.pull.body.toLowerCase().indexOf("close");
+    // Judge PR is associated with an Issue and uses close semantics
+    if (issueNumber && closeIndex != -1) {
+      const issue = await this.issueRepository.findOne({
+        issueNumber: issueNumber,
+      });
+      const pull = await this.pullRepository.findOne({
+        pullNumber: pullPayload.pull.number,
+      });
+      if (issue == undefined || pull == undefined) {
```
Comment (hi-rustin) [inline src/services/challenge-pull/index.ts:402]: Use `====` replace `==`

File: src/services/challenge-pull/index.ts [hunk 4]
```diff
@@ -385,4 +385,44 @@ export default class ChallengePullService implements IChallengePullService {
       message: ChallengePullMessage.Rewarded,
     };
   }
+
+  public async awardWhenPullClosedAndContainClose(
+    pullPayload: PullPayload
+  ): Promise<boolean> {
+    const issueNumber = findLinkedIssueNumber(pullPayload.pull.body);
+    const closeIndex = pullPayload.pull.body.toLowerCase().indexOf("close");
```
Comment (hi-rustin) [inline src/services/challenge-pull/index.ts:393]: We should match Issue Number: close xxx as a whole, not the close keyword directly.

Comment (whileTrue0x90) [reply src/services/challenge-pull/index.ts]: I use re to solve this problem, please help me check the code again

Comment (hi-rustin) [reply src/services/challenge-pull/index.ts]: You can use https://github.com/ti-community-infra/ti-challenge-bot/blob/2377b98b34c0eeeebfae7f1b930d11b2c8fde6c6/src/services/utils/PullUtil.ts#L4.

Comment (whileTrue0x90) [reply src/services/challenge-pull/index.ts]: After thinking about it, I think we should build a new method

File: src/commands/reward/index.ts [hunk 5]
```diff
@@ -69,9 +69,10 @@ const reward = async (
   }
 
   // Find linked issue assignees.
-  const issueNumber = findLinkedIssueNumber(data.body);
+  const issueMsg = findLinkedIssueNumber(data.body);
+  const issueNumber = issueMsg[0];
```
Comment (hi-rustin) [inline src/commands/reward/index.ts:73]: I think you can give it a name. Not directly use the index.

File: src/commands/reward/index.ts [hunk 6]
```diff
@@ -69,9 +69,10 @@ const reward = async (
...
 
-  if (issueNumber === null) {
+  if (issueNumber === null || typeof issueNumber === "boolean") {
```
Comment (hi-rustin) [inline src/commands/reward/index.ts:75]: Why we need this type of?

**Review by hi-rustin — CHANGES_REQUESTED** [review]

File: src/commands/reward/index.ts (hunk 6 above)
Comment (whileTrue0x90) [reply src/commands/reward/index.ts:75]: I improved the method findlinkedussuenumber, which returns null [] | (number | Boolean) []

File: src/commands/reward/index.ts (hunk 5 above)
Comment (whileTrue0x90) [reply src/commands/reward/index.ts:73]: Sorry, I didn't understand. Issue MSG is a tuple。I want to get the first element

Comment (hi-rustin) [reply src/commands/reward/index.ts:73]: const [firstName,secondName] = findLinkedIssueNumber(data.body);

Comment (whileTrue0x90) [reply src/commands/reward/index.ts:73]: ok, I understand.

Comment (whileTrue0x90) [reply src/commands/reward/index.ts:73]: I'm worried. secondName, I didn't use it.

Comment (hi-rustin) [reply src/commands/reward/index.ts:73]: You can ignore the second value.

Comment (whileTrue0x90) [reply src/commands/reward/index.ts]: I solved the problem

**Pull request closed without merging**